src/main/java/org/folio/circulation/resources/LostItemsRequiringActualCostResource.java

```diff
@@ -0,0 +1,204 @@
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+
+import org.folio.circulation.domain.Account;
+import org.folio.circulation.domain.FeeFine;
+import org.folio.circulation.domain.FeeFineOwner;
+import org.folio.circulation.domain.Item;
+import org.folio.circulation.domain.LostItemRequiringCostsFee;
+import org.folio.circulation.domain.Loan;
+import org.folio.circulation.domain.LoanStatus;
+import org.folio.circulation.domain.MultipleRecords;
+import org.folio.circulation.infrastructure.storage.ServicePointRepository;
+import org.folio.circulation.infrastructure.storage.feesandfines.FeeFineOwnerRepository;
+import org.folio.circulation.infrastructure.storage.inventory.ItemRepository;
+import org.folio.circulation.infrastructure.storage.loans.LoanRepository;
+import org.folio.circulation.infrastructure.storage.loans.LostItemPolicyRepository;
+import org.folio.circulation.infrastructure.storage.users.UserRepository;
+import org.folio.circulation.support.Clients;
+import org.folio.circulation.support.RouteRegistration;
+import org.folio.circulation.support.http.client.CqlQuery;
+import org.folio.circulation.support.http.client.Offset;
+import org.folio.circulation.support.http.client.PageLimit;
+import org.folio.circulation.support.http.server.JsonHttpResponse;
+import org.folio.circulation.support.http.server.WebContext;
+import org.folio.circulation.support.results.Result;
+
+import io.vertx.core.http.HttpClient;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.Router;
+import io.vertx.ext.web.RoutingContext;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+public class LostItemsRequiringActualCostResource extends Resource {
+  private static final String RECORD_KEY = "lostItems";
+  private static final String TOTAL_RECORDS_KEY = "totalRecords";
+
+  private final String rootPath;
+
+  public LostItemsRequiringActualCostResource(String rootPath, HttpClient client) {
+    super(client);
+    this.rootPath = rootPath;
+  }
+
+  @Override
+  public void register(Router router) {
+    RouteRegistration routeRegistration = new RouteRegistration(rootPath, router);
+    routeRegistration.getMany(this::getMany);
+  }
+
+  private void getMany(RoutingContext routingContext) {
+    final WebContext context = new WebContext(routingContext);
+    final Clients clients = Clients.create(context, client);
+    final Repositories repositories = new Repositories(clients);
+    final PageLimit pageLimit = PageLimit.limit(context.getIntegerParameter("limit", 100));
+    final Offset offset = Offset.offset(context.getIntegerParameter("offset", 0));
+
+    repositories.getLoan().getMany(buildQuery().value(), pageLimit, offset)
```

Comment:
How does paging work when the records are filtered after the storage query for loans?

If some of these records are filtered out, won't that lead to partial pages and subsequent pages starting at the incorrect point.

Reply:
I was worried about the same thing.

This either must be fully handled by SQL (with table joins etc...which won't work with CQL as far as I am aware) or we need to handle pagination outside of the database. Which does not sound like a good idea to me.

Would it be better to have **no** pagination and offset, in which case it would just return everything?

Should I instead, use the above queries to get a list of all IDs and then load by IDs with pagination and offset?
I imagine this would effectively go through everything potentially 2 times and the list of IDs could be large.

Reply:
> This either must be fully handled by SQL (with table joins etc...which won't work with CQL as far as I am aware) or we need to handle pagination outside of the database. Which does not sound like a good idea to me.

This isn't a viable option in FOLIO at the moment, because items and loans come from different modules.

I think trying to support paging outside of the database likely means an overall heavier burden on the system, as each subsequent page becomes more expensive to determine.

> Would it be better to have **no** pagination and offset, in which case it would just return everything?

From my perspective, I think so, as I don't know how the client benefits from this.

I wasn't present for the conversation where it was decided this would be a just in time report.

I imagine that folks want to review the whole report at once. Given that, I don't know what paging from the clients perspective would achieve.

> Should I instead, use the above queries to get a list of all IDs and then load by IDs with pagination and offset?
> I imagine this would effectively go through everything potentially 2 times and the list of IDs could be large.

I had thought that the decision to do this as a just-in-time report was based upon the size of the report being small.

I would prefer not to need to process the loans twice if possible.

I think there may be value in internally paging the loans, in order to limit memory usage in mod-circulation.

That would rely on some kind of accumulation that meant that most of the information uses to process those loans was thrown away before the next page is fetched.

@bohdan-suprun Did you recently implement something like this, in order to reduce high memory in another report?

There is a general challenge in mod-circulation with fetching large sets of related records. We cannot use the database to JSON streaming techniques that some of the storage APIs use to circumvent this because the output JSON may need multiple source records.

Over the longer term, there is an argument for changing the way related records are fetched. Either to do so individually, which might make a (what FOLIO calls) streaming approach more viable at the cost of many more API requests or (my preference) for FOLIO to accept the use of secondary representations and instead have circulation keep a local representation of related records. Alas, the latter approach would encounter some challenges with FOLIO's current architecture (and process design in some cases).

Reply:
There is `PageableFetcher`, it needs to be adopted a little bit for this particular case. If we consider using the fetcher then we probably should introduce a report entry class, so when we fetched a page of loans, we should join all reference records needed and then immediately convert these records to report entries, so that much less objects will be kept in memory.

Ideal solution will be if we stream response to the client from the pageable fetcher (vert.x call it _chunked_ response - every `write` call on http response is sent as separate HTTP response), in this case amount of objects that are in memory is minimal.

Reply:
Of course, there might be some common challenges that called phantom reads, for example:
1. we fetch open loans and order them by dueDate;
1. Let's say we've already fetched 2 pages and the latest due date was 2020-11-20;
1. Then, a user checked out a book and short term loan policy applied, so a loan created and dueDate calculated to 2020-11-11 8 PM;
1. In this case this newly created loan should be somewhere on the first 2 pages, but we already read them, so this loan won't be processed.

But I hope it is not a critical issue for this particular case, and most of the pageable API endpoints have this issue.